File: voom-fe/src/app/shared/rest/user/user.api.ts
import { Api } from '../api';
import { RequestConfig } from '../rest.model';
import { CreateUserDto, UserProfileDto } from './user.model';
import { ApiClient } from '../api-client';
import { AdminCreateDriverDto, DriverDto } from '../driver/driver.model';

export interface BlockUserRequestDto {
  reason: string;
}

export interface UserBlockNoteDto {
  id: number;
  userId: number;
  reason: string;
  createdAt: string;
  adminId: number;
  active: boolean;
}

export class UserApi extends Api {
  constructor(apiClient: ApiClient) {
    super(apiClient);
  }

  getUsers() {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.get<void, UserProfileDto[]>('/api/users', config);
  }

  getUser(id: number) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.get<void, UserProfileDto>(`/api/users/${id}`, config);
  }

  createUser(dto: CreateUserDto) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.post<CreateUserDto, UserProfileDto>(`/api/users`, dto, config);
  }

  updateUser(id: number, userDto: UserProfileDto) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.put<UserProfileDto, UserProfileDto>(`/api/users/${id}`, userDto, config);
  }

  deleteUser(id: number) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.delete<void, void>(`/api/users/${id}`, config);
  }

  blockUser(id: number, dto: { reason: string }) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.post<{ reason: string }, UserProfileDto>(
      `/api/users/${id}/block`,
      dto,
      config,
    );
  }

  unblockUser(id: number) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.post<void, UserProfileDto>(`/api/users/${id}/unblock`, undefined, config);
  }

  getActiveBlockNote(userId: number) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.get<void, UserBlockNoteDto>(`/api/users/${userId}/block-note`, config);
  }
}

File: voom-app/app/src/main/java/inc/visor/voom/app/shared/api/UserApi.java
package inc.visor.voom.app.shared.api;

import java.util.List;

import inc.visor.voom.app.shared.dto.user.CreateUserDto;
import inc.visor.voom.app.shared.dto.user.UserProfileDto;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;

public interface UserApi {

    @GET("/api/users")
    Call<List<UserProfileDto>> getUsers();

    @GET("/api/users/{userId}")
    Call<UserProfileDto> getUser(@Path("userId") Long userId);

    @POST("/api/users")
    Call<UserProfileDto> createUser(@Body CreateUserDto dto);

    @PUT("/api/users/{userId}")
    Call<UserProfileDto> updateUser(
            @Path("userId") Long userId,
            @Body UserProfileDto dto
    );

    @DELETE("/api/users/{userId}")
    Call<Void> deleteUser(@Path("userId") Long userId);

}
